File: Source/Aladdin.CAPI.GOST.Java/src/aladdin/capi/gost/keyx/gostr3410/TransportKeyWrap.java
package aladdin.capi.gost.keyx.gostr3410;
import aladdin.*; 
import aladdin.asn1.*;
import aladdin.asn1.iso.*;
import aladdin.asn1.gost.*;
import aladdin.capi.*; 
import java.security.*;
import java.io.*;

////////////////////////////////////////////////////////////////////////////
// Согласование ключа на стороне-отправителе
////////////////////////////////////////////////////////////////////////////
public class TransportKeyWrap extends aladdin.capi.TransportKeyWrap
{
    // фабрика алгоритмов и область видимости
    private final Factory factory; private final SecurityStore scope; 
    // идентификатор алгоритма согласования ключа
    private final String transportAgreementOID;     
    
    // конструктор
    public TransportKeyWrap(Factory factory, SecurityStore scope, String transportAgreementOID) 
    {  
        // сохранить переданные параметры
        this.factory = RefObject.addRef(factory);
        this.scope   = RefObject.addRef(scope  ); 
        
        // сохранить переданные параметры
        this.transportAgreementOID = transportAgreementOID;
    }
    // освободить выделенные ресурсы
    @Override protected void onClose() throws IOException   
    {
        // освободить выделенные ресурсы
        RefObject.release(scope); RefObject.release(factory); super.onClose();
    }
    // зашифровать ключ
    @Override public TransportKeyData wrap(AlgorithmIdentifier algorithmParameters, 
        IPublicKey publicKey, IRand rand, ISecretKey CEK) throws IOException, InvalidKeyException 
    {
        // преобразовать тип параметров
        aladdin.capi.gost.gostr3410.INamedParameters parameters = 
            (aladdin.capi.gost.gostr3410.INamedParameters)publicKey.parameters(); 
        
        // создать алгоритм согласования ключа (ESDH)
        try (ITransportAgreement transportAgreement = 
            createAgreementAlgorithm(factory, scope, parameters.sboxOID()))
       { 
            // зашифровать ключ
            TransportAgreementData agreementData = transportAgreement.wrap(
                null, null, new IPublicKey[] { publicKey }, rand, CEK
            ); 
            // раскодировать зашифрованный ключ
            EncryptedKey sessionEncryptedKey = new EncryptedKey(
                Encodable.decode(agreementData.encryptedKeys[0])
            ); 
            // закодировать параметры и содержимое ключа 
            aladdin.asn1.iso.pkix.SubjectPublicKeyInfo ephemeralPublicKey = 
                agreementData.publicKey.encoded(); 
                
            // закодировать параметры транспортировки
            GOSTR3410TransportParameters transportParameters = new GOSTR3410TransportParameters(
                new ObjectIdentifier(parameters.sboxOID()), 
                ephemeralPublicKey, new OctetString(agreementData.random)
            ); 
            // закодировать зашифрованный ключ с параметрами
            GOSTR3410KeyTransport encodedEncryptedKey = new GOSTR3410KeyTransport(
                sessionEncryptedKey, transportParameters
            );
            // вернуть параметры обмена ключа и зашифрованный ключ
            return new TransportKeyData(
                ephemeralPublicKey.algorithm(), encodedEncryptedKey.encoded()
            ); 
        }
    }
    protected ITransportAgreement createAgreementAlgorithm(
        Factory factory, SecurityStore scope, String sboxOID) throws IOException
    {
        // указать идентификатор алгоритма
        AlgorithmIdentifier wrapParameters = new AlgorithmIdentifier(
            new ObjectIdentifier(OID.KEY_WRAP_CRYPTOPRO), 
            new KeyWrapParameters(new ObjectIdentifier(sboxOID), null)
        );
        // указать идентификатор алгоритма
        AlgorithmIdentifier transportParameters = new AlgorithmIdentifier(
            new ObjectIdentifier(transportAgreementOID), wrapParameters
        ); 
        // создать алгоритм согласования ключа
        try (ITransportAgreement transportAgreement = (ITransportAgreement)
            factory.createAlgorithm(
                scope, transportParameters, ITransportAgreement.class))
        {
            // проверить поддержку алгоритмов
            if (transportAgreement == null) throw new UnsupportedOperationException(); 
        
            // для алгоритма ESDH
            if (transportAgreement instanceof aladdin.capi.keyx.ESDH) 
            {
                // вернуть алгоритм
                return RefObject.addRef(transportAgreement); 
            }
            // создать алгоритм ESDH
            return new aladdin.capi.keyx.ESDH(factory, transportAgreement); 
        }
    }
}

File: Source/Aladdin.CAPI.GOST/CAPI/GOST/Keyx/GOSTR3410/TransportKeyWrap.cs
using System; 

namespace Aladdin.CAPI.GOST.Keyx.GOSTR3410
{
    ////////////////////////////////////////////////////////////////////////////
    // Согласование ключа на стороне-отправителе ГОСТ Р 34.10
    ////////////////////////////////////////////////////////////////////////////
    public class TransportKeyWrap : CAPI.TransportKeyWrap
    {
        // фабрика алгоритмов и область видимости
        private CAPI.Factory factory; private SecurityStore scope;
        // идентификатор алгоритма
        private string transportAgreementOID;     

        // конструктор
        public TransportKeyWrap(CAPI.Factory factory, SecurityStore scope, string transportAgreementOID)
        {
            // сохранить переданные параметры
            this.factory = RefObject.AddRef(factory); 
            this.scope   = RefObject.AddRef(scope  ); 
            
            // сохранить переданные параметры
            this.transportAgreementOID = transportAgreementOID; 
        }
        // освободить выделенные ресурсы
        protected override void OnDispose()
        {
            // освободить выделенные ресурсы
            RefObject.Release(scope); RefObject.Release(factory); base.OnDispose();
        }
        // зашифровать ключ
        public override TransportKeyData Wrap(
            ASN1.ISO.AlgorithmIdentifier algorithmParameters, 
            CAPI.IPublicKey publicKey, IRand rand, ISecretKey CEK)
        {
            // преобразовать тип параметров
            GOST.GOSTR3410.INamedParameters parameters =
                (GOST.GOSTR3410.INamedParameters)publicKey.Parameters;

            // создать алгоритм согласования ключа (ESDH)
            using (ITransportAgreement transportAgreement = CreateAgreementAlgorithm(
                factory, scope, parameters.SBoxOID))
            { 
                // зашифровать ключ
                TransportAgreementData agreementData = transportAgreement.Wrap(
                    null, null, new IPublicKey[] { publicKey }, rand, CEK
                ); 
                // раскодировать зашифрованный ключ
                ASN1.GOST.EncryptedKey sessionEncryptedKey =
                    new ASN1.GOST.EncryptedKey(
                        ASN1.Encodable.Decode(agreementData.EncryptedKeys[0]
                ));
                // закодировать параметры и содержимое ключа 
                ASN1.ISO.PKIX.SubjectPublicKeyInfo ephemeralPublicKey =
                    agreementData.PublicKey.Encoded;

                // закодировать параметры транспортировки
                ASN1.GOST.GOSTR3410TransportParameters transportParameters =
                    new ASN1.GOST.GOSTR3410TransportParameters(
                        new ASN1.ObjectIdentifier(parameters.SBoxOID),
                        ephemeralPublicKey, new ASN1.OctetString(agreementData.Random)
                );
                // закодировать зашифрованный ключ с параметрами
                ASN1.GOST.GOSTR3410KeyTransport encodedEncryptedKey =
                    new ASN1.GOST.GOSTR3410KeyTransport(sessionEncryptedKey, transportParameters);

                // вернуть зашифрованный ключ
                return new TransportKeyData(ephemeralPublicKey.Algorithm, encodedEncryptedKey.Encoded);
            }                    
        }
        protected virtual ITransportAgreement CreateAgreementAlgorithm(
            CAPI.Factory factory, SecurityStore scope, string sboxOID)
        {
            // указать параметры алгоритма
            ASN1.GOST.KeyWrapParameters keyWrapParameters = new ASN1.GOST.KeyWrapParameters(
                new ASN1.ObjectIdentifier(sboxOID), null
            ); 
            // указать идентификатор алгоритма
            ASN1.ISO.AlgorithmIdentifier wrapParameters = new ASN1.ISO.AlgorithmIdentifier(
                new ASN1.ObjectIdentifier(ASN1.GOST.OID.keyWrap_cryptopro), keyWrapParameters
            );
            // указать идентификатор алгоритма
            ASN1.ISO.AlgorithmIdentifier transportParameters = new ASN1.ISO.AlgorithmIdentifier(
                new ASN1.ObjectIdentifier(transportAgreementOID), wrapParameters
            ); 
            // создать алгоритм согласования ключа
            using (ITransportAgreement transportAgreement = 
                factory.CreateAlgorithm<ITransportAgreement>(scope, transportParameters))
            { 
                // проверить поддержку алгоритмов
                if (transportAgreement == null) throw new NotSupportedException(); 
            
                // проверить тип алгоритма
                if (transportAgreement is CAPI.Keyx.ESDH) return RefObject.AddRef(transportAgreement); 
            
                // создать алгоритм ESDH
                return new CAPI.Keyx.ESDH(factory, transportAgreement); 
            }
        }
    }
}
